### PR TITLE
feature cache

### PR DESCRIPTION
**Pull request by JanDeDobbeleer**
- refactor: split concurrent map from cache implementation
- feat: cache capabilities

### Prerequisites

- [x] I have read and understood the `CONTRIBUTING` guide
- [x] The commit message follows the [conventional commits][cc] guidelines
- [x] Tests for the changes have been added (for bug fixes/features)
- [x] Docs have been added / updated (for bug fixes/features)

### Description

[Description of the change]

[cc]: https://www.conventionalcommits.org/en/v1.0.0/#summary

### CODE REVIEW

**Comment by lnu** [comment]
I did something similar on my side today… 
With load/save from/to a file

File: src/environment_cache.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+func (fc *fileCache) init(home string) {
+	fc.cache = newConcurrentMap()
+	fc.home = home
+	content, err := ioutil.ReadFile(home + cachePath)
+	if err != nil {
+		return
+	}
+	for _, line := range strings.Split(string(content), "\n") {
```
Comment (lnu) [inline src/environment_cache.go]: I did serialize using  gob.NewEncoder/gob.NewDecoder but yours can be easily opened with a text editor. The only missing part is an expiration timeout.

File: src/environment_cache.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+func (fc *fileCache) close() {
+	if dump, err := json.Marshal(fc.cache.list()); err == nil {
+		_ = ioutil.WriteFile(fc.home+cachePath, dump, 0644)
+	}
+}
+
+func (fc *fileCache) get(key string) (*cacheObject, bool) {
+	val, found := fc.cache.get(key)
```
Comment (lnu) [inline src/environment_cache.go:49]: Maybe checking the timestamp before returning the value from the cache

File: src/environment_cache.go (hunk 1 above)
Comment (JanDeDobbeleer) [reply src/environment_cache.go]: Added, but I'll refactor so you set the TTL in seconds when adding. A lot easier for consumption/handling.

File: src/environment_cache.go (hunk 2 above)
Comment (JanDeDobbeleer) [reply src/environment_cache.go:49]: Yes. That's the idea now that I wrote this. That should be internal and not leak.

**Comment by JanDeDobbeleer** [comment]
> I did something similar on my side today…
> With load/save from/to a file

@lnu I'd appreciate a review, if architecturally OK I can start adding some tests

File: src/environment.go [hunk 3]
```diff
@@ -77,24 +85,24 @@ type environmentInfo interface {
...
 func (c *commandCache) get(command string) (string, bool) {
-	c.lock.RLock()
-	defer c.lock.RUnlock()
-	if cmd, ok := c.commands[command]; ok {
-		command = cmd
+	cmd, found := c.commands.get(command)
+	if !found {
+		return "", false
+	}
+	if command, ok := cmd.(string); ok {
```
Comment (JanDeDobbeleer) [inline src/environment.go]: remove this and simply use `return cmd.(string)`

File: src/environment_cache.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+	}
+
+	for key, co := range list {
+		fc.cache.set(key, co)
+	}
+}
+
+func (fc *fileCache) close() {
+	fc.set("hello", "world", 200)
+	if dump, err := json.Marshal(fc.cache.list()); err == nil {
```
Comment (JanDeDobbeleer) [inline src/environment_cache.go]: format the output for human readability

Comment (lnu) [reply src/environment_cache.go]: Is it possible to format it for human readbility with the addition of the timestamp?

Comment (lnu) [reply src/environment_cache.go]: On my side it was serialized and thus impossible to read for a "human". Maybe we can add an extra parameter to omp to clear or display the content of the cache?

File: src/environment.go (hunk 3 above)
Comment (lnu) [reply src/environment.go]: I did the same, returning the value with a bool to tell if found or not. It's easier to use after for the caller.

File: src/main.go [hunk 5]
```diff
@@ -160,7 +160,7 @@ func main() {
 	flag.Parse()
 	env := &environment{}
 	env.init(args)
-	defer env.tracer.close()
+	defer env.close()
```
Comment (lnu) [inline src/main.go:163]: Good idea to have a global close

**Comment by lnu** [comment]
All the code I've done is here https://github.com/boarder2/oh-my-posh3/pull/1/files
It's very rough but it's very similar to yours
I still wonder how to return a mocked cache from the mocked environment.

**Pull request merged**